src/miner_lora_throttle.erl

```diff
@@ -0,0 +1,241 @@
+    Pred = fun (Sent) -> Sent#sent_packet.sent_at > CutoffTime end,
+    lists:takewhile(Pred, SentPackets);
+trim_sent('EU868', SentPackets = [H | _]) ->
+    CutoffTime = H#sent_packet.sent_at - ?DUTY_CYCLE_PERIOD_MS,
+    Pred = fun (Sent) -> Sent#sent_packet.sent_at > CutoffTime end,
+    lists:takewhile(Pred, SentPackets).
+
+%% @doc Based on previously sent packets, returns a boolean value if
+%% it is legal to send on Frequency at time Now.
+%%
+%%
+-spec can_send(
+    Handle :: handle(),
+    AtTime :: number(),
+    Frequency :: integer(),
+    TimeOnAir :: number()
+) ->
+    boolean().
+can_send(_Handle, _AtTime, _Frequency, TimeOnAir) when TimeOnAir > ?MAX_DWELL_TIME_MS ->
+    %% TODO: double check that ETSI's max time on air is the same as
+    %% FCC.
+    false;
+can_send({undefined, _SentPackets}, _AtTime, _Frequency, _TimeOnAir) ->
+    false;
+can_send({'US915', SentPackets}, AtTime, Frequency, TimeOnAir) ->
+    CutoffTime = AtTime - ?DWELL_TIME_PERIOD_MS + TimeOnAir,
+    ProjectedDwellTime = dwell_time(SentPackets, CutoffTime, Frequency) + TimeOnAir,
+    ProjectedDwellTime =< ?MAX_DWELL_TIME_MS;
+can_send({'EU868', SentPackets}, AtTime, Frequency, TimeOnAir) ->
+    CutoffTime = AtTime - ?DUTY_CYCLE_PERIOD_MS,
+    CurrDwell = dwell_time(SentPackets, CutoffTime, Frequency),
+    OnePercent = 0.01,
+    (CurrDwell + TimeOnAir) / ?DUTY_CYCLE_PERIOD_MS < OnePercent.
+
+%% @doc Computes the total time on air for packets sent on Frequency
+%% and no older than CutoffTime.
+-spec dwell_time(list(#sent_packet{}), integer(), number()) -> number().
+dwell_time(SentPackets, CutoffTime, Frequency) ->
+    dwell_time(SentPackets, CutoffTime, Frequency, 0).
+
+-spec dwell_time(list(#sent_packet{}), integer(), number(), number()) -> number().
+%% Scenario 1: entire packet sent before CutoffTime
+dwell_time([P | T], CutoffTime, Frequency, Acc)
+        when P#sent_packet.sent_at + P#sent_packet.time_on_air < CutoffTime ->
+    dwell_time(T, CutoffTime, Frequency, Acc);
+%% Scenario 2: packet sent on non-relevant frequency.
+dwell_time([P | T], CutoffTime, Frequency, Acc) when P#sent_packet.frequency /= Frequency ->
+    dwell_time(T, CutoffTime, Frequency, Acc);
+%% Scenario 3: Packet started before CutoffTime but finished after CutoffTime.
+dwell_time([P | T], CutoffTime, Frequency, Acc) when P#sent_packet.sent_at =< CutoffTime ->
+    RelevantTimeOnAir = P#sent_packet.time_on_air - (CutoffTime - P#sent_packet.sent_at),
+    true = RelevantTimeOnAir >= 0,
+    dwell_time(T, CutoffTime, Frequency, Acc + RelevantTimeOnAir);
+%% Scenario 4: 100 % of packet transmission after CutoffTime.
+dwell_time([P | T], CutoffTime, Frequency, Acc) ->
+    dwell_time(T, CutoffTime, Frequency, Acc + P#sent_packet.time_on_air);
+dwell_time([], _CutoffTime, _Frequency, Acc) ->
+    Acc.
+
+%% @doc Returns total time on air for packet sent with given
+%% parameters.
+%%
+%% See Semtech Appnote AN1200.13, "LoRa Modem Designer's Guide"
```

Comment:
It'd be nice if you included a short summary of the note below this comment

Reply:
I'd be happy to, but I'm not sure what what there is to summarize. The function body is cut & paste from the app note and erlangified.

Reply:
Maybe just include a link?